src/Sylius/Bundle/ResourceBundle/Doctrine/ODM/PHPCR/EventListener/NameResolverListener.php

```diff
@@ -0,0 +1,105 @@
+<?php
+
+namespace Sylius\Bundle\ResourceBundle\Doctrine\ODM\PHPCR\EventListener;
+
+use Symfony\Component\EventDispatcher\EventSubscriberInterface;
+use Symfony\Component\Form\FormEvents;
+use Symfony\Component\Form\FormEvent;
+use Doctrine\ODM\PHPCR\DocumentManagerInterface;
+use PHPCR\Util\NodeHelper;
+use Doctrine\ODM\PHPCR\Mapping\ClassMetadata;
+use Sylius\Bundle\ResourceBundle\Event\ResourceControllerEvent;
+
+/**
+ * Handles the resolution of the PHPCR node name field.
+ *
+ * - If a node already exists with the same name, then a numerical index will
+ *   be appended to the name.
+ * - Invalid characters will be replaced with "_"
+ * 
+ * @see http://www.day.com/specs/jcr/2.0/3_Repository_Model.html#3.2.2%20Local%20Names
+ * @see https://github.com/phpcr/phpcr-utils/blob/master/src/PHPCR/Util/PathHelper.php#L95
+ *
+ * TODO: Allow specification of replacement character?
+ * TODO: Allow strict mode? i.e. throw exceptions if the node exists or if it
+ *       contains illegal characters (in which case a Validator should also be
+ *       added).
+ *
+ * @author Daniel Leech <daniel@dantleech.com>
+ */
+class NameResolverListener
+{
+    const REPLACEMENT_CHAR = ' ';
+
+    /**
+     * @var DocumentManagerInterface
+     */
+    private $documentManager;
+
+    /**
+     * @param DocumentManagerInterface $documentManager
+     */
+    public function __construct(
+        DocumentManagerInterface $documentManager
+    )
+    {
+        $this->documentManager = $documentManager;
+    }
+
+
+    public function onPreCreate(ResourceControllerEvent $event)
+    {
+        $document = $event->getSubject();
+
+        if (!is_object($document)) {
+            throw new \InvalidArgumentException(sprintf(
+                'Expected an object, got a "%s"',
+                gettype($document)
+            ));
+        }
+
+        $metadata = $this->documentManager->getClassMetadata(get_class($document));
+
+        if ($metadata->idGenerator !== ClassMetadata::GENERATOR_TYPE_PARENT) {
+            return;
+        }
+
+        // TODO: Throw exception in these two cases? PHPCR-ODM will fail
+        //       (later) anyway if either of these are true.
+        if (null === $nameField = $metadata->nodename) {
```

Comment:
User has already explicitly declared that they want to use a default path by this stage, should probably throw Exceptions.